File: src/042/p042.py
import math
import os.path


LETTER_SCORES = {chr(ord("A") + i): i + 1 for i in range(26)}


def get_word_score(word: str) -> int:
    return sum(LETTER_SCORES[char] for char in word)


def is_triangle_number(number: int) -> bool:
    return math.sqrt(8 * number + 1) % 1 == 0


def solve() -> int:
    input_path = os.path.join(os.path.dirname(__file__), "p042_words.txt")
    with open(input_path, encoding="utf8") as file:
        content = file.read()
    words = sorted(word.strip('"') for word in content.split(","))
    return sum(1 for w in words if is_triangle_number(get_word_score(w)))

File: src/042/p042.go
package problem042

import (
	"embed"
	"strconv"
	"strings"

	"github.com/sunilbpandey/project-euler/src/utils/go/errorutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/math/intutils"
)

//go:embed p042_words.txt
var content embed.FS

func getWordScore(word string) int {
	score := 0
	for _, c := range word {
		score += int(c) - 'A' + 1
	}
	return score
}

func isPerfectSquare(n int) bool {
	sqrt := intutils.Sqrt(n)
	return sqrt*sqrt == n
}

func Solve() string {
	content, err := content.ReadFile("p042_words.txt")
	errorutils.Check(err)

	count := 0
	words := strings.Split(string(content), ",")
	for _, word := range words {
		word = strings.Trim(word, "\"")
		score := getWordScore(word)
		if isPerfectSquare(8*score + 1) {
			count++
		}
	}
	return strconv.Itoa(count)
}
